ships/0024-surfacing-errors-to-buildrun.md

```diff
@@ -0,0 +1,187 @@
+
+Additionally, the proposal enables a user to keep the troubleshooting
+process on the BuildRun level and only descend further into Tekton
+specifics when necessary.
+
+### Goals
+
+- A mechanism to automatically pick up error results from a TaskRun
+  Surface them to the associated BuildRun
+
+### Non-Goals
+
+- The specifics of errors in regards to tooling are not part of this
+  proposal. Neither is their means for reporting them. Errors highly
+  depend on each strategy
+
+## Proposal
+
+We propose to introduce two generic results for errors:
+`shp-error-message` and `shp-error-reason`.  The result
+`shp-error-reason` have a recognizable name that is ideally prefixed
+with the problematic technology, for example, `git` in case of missing
+a secret. The result `shp-error-message` should be descriptive text to
+give the user enough information to troubleshoot the BuildRun.  A task
+run emits the new results as Tekton results.  The BuildRun CRD is
+assumed to have the `status.results` according to `0023` (alternatives
+are discussed further below). 
+Finally, we must extend the BuildRun controller
+logic to propagate the new results into the BuildRun CRD's
+`status.results` property.
+
+BuildRun with new error results: 
+
+```yaml
+status:
+   results:
+       - name: shp-error-reason
+         value: git-ssh-auth
+       - name: shp-error-message
+         value: Authentication failed for ...
+```
+
+We do not suggest to mutate the `status.conditions.message` field
+since it contains valuable hints to troubleshoot on a lower level if
+the propagated error message and reason is not sufficient.
+
+### User Stories
+
+#### Story 1 - Developer and direct user of shipwright perspective
+As a user of Shipwright and developer, I want to identify why my
+BuildRun is in a failed state without going through the lengthy
+process of identifying the correct container and manually processing
+its logs.
+#### Story 2 - Github Actions developer (third party application developer)
+As a third-party application developer that uses shipwright under the hood, I want to supply my users with meaningful hints to fix their builds, thus requiring a straightforward API to consume error details.
+
+### Implementation Notes
+
+`tbd in the next phase of the proposal`
+
+### Test Plan
+
+`tbd in the next phase of the proposal`
+
+### Release Criteria
+
+`tbd in the next phase of the proposal`
+
+### Risks and Mitigations
+Tekton results are limited in size (4096 bytes) which applies to the
+proposed error results.
+
+There is no strict typing that would prevent strategy authors from
+using proper error reasons. So two authors can use different names for
+the same error reason or use the same name for different errors. To
+prevent these issues, we need clear documentation with well-known
+error reasons.
```

Comment:
I do not see that so critical. Important is that we document the errors that we write, for example from the git step. The author of a build strategy then must define what his errors are.

Reply:
My thoughts were that a third party application can have specific error handling logic for dealing with failed buildruns. For example a UI that asks the user to reupload their ssh key. No strict typing makes it less reliable (there is no build time guarantee for consistent error reasons). 

I can remove the information from that section since it is not so critical.

Reply:
I think one will have to separate two kinds of client applications (like UIs) here:

- Generic clients like `shp` CLI that do not know the build strategies. They will not be able to provide much support on those errors other than showing the message which is meant to be for the end user. The reason they can show but that's not so much of relevance for them.
- Specific clients that know the build strategies (like in Code Engine). We can also make them know the error reasons that the build strategies return and put in approprioate error handling (like localized error messages) that go beyond the message from the build strategy. In this case, the build strategies are also coming all from the same source (us) and one can make sure that same reasons are used in different strategies for the same root cause.

Reply:
I have adjusted the sentence to make its intent more clear.